lib/ast/expression.ts

```diff
@@ -1191,3 +1191,88 @@ export class ChainExpression extends Expression {
         return [null, newScope];
     }
 }
+
+
+export class JoinExpression extends Expression {
+    lhs : Expression;
+    rhs : Expression;
+    condition : BooleanExpression|null;
+
+    constructor(location : SourceRange|null,
+                left : Expression,
+                right : Expression,
+                condition : BooleanExpression|null, 
+                schema : FunctionDef|null) {
+        super(location, schema);
+
+        this.lhs = left;
+        this.rhs = right;
+        this.condition = condition;
+    }
+
+    get priority() : SyntaxPriority {
+        return SyntaxPriority.Join;
+    }
+
+    toSource() : TokenStream {
+        if (this.condition) {
+            return List.concat(
+                addParenthesis(this.priority, this.lhs.priority, this.lhs.toSource()),
+                'join',
+                addParenthesis(this.priority, this.rhs.priority, this.rhs.toSource()),
+                'on',
+                addParenthesis(this.priority, this.condition.priority, this.condition.toSource())
+            );
+        } else {
+            return List.concat(
+                addParenthesis(this.priority, this.lhs.priority, this.lhs.toSource()),
+                'join',
+                addParenthesis(this.priority, this.rhs.priority, this.rhs.toSource())
+            );
+        }
+    }
+
+    equals(other : Expression) : boolean {
+        return other instanceof JoinExpression &&
+            this.lhs === other.lhs && this.rhs === other.rhs && this.condition === other.condition;
```

Comment:
All the === needs to be .equals() instead.

Reply:
fixed in 61a6e46.